project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java

```diff
@@ -582,78 +454,126 @@ private List<MeasurementId> performRegistrationPxp(
       }
     }
     Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> proteomicsMeasurementsMapping = new HashMap<>();
-    for (ProteomicsMeasurementMetadata metadata : proteomicsMeasurements) {
-      proteomicsMeasurementsMapping.putAll(preparePxpMeasurement(projectId, metadata));
-    }
+
+    // Start with the pooled measurements first and group the metadata entries by pool
+    Map<String, List<ProteomicsMeasurementMetadata>> measurementsByPool = proteomicsMeasurements.stream()
+        .filter(metadata -> metadata.assignedSamplePoolGroup().isPresent())
+        .collect(Collectors.groupingBy(metadata -> metadata.assignedSamplePoolGroup().get()));
+
+    // We collect the "single" sample measurements extra
+    List<ProteomicsMeasurementMetadata> singleMeasurements = proteomicsMeasurements.stream()
+        .filter(metadata -> metadata.assignedSamplePoolGroup().isEmpty()).toList();
+
+    // Then merge and prepare the domain objects by pool
+    proteomicsMeasurementsMapping.putAll(mergeByPool(measurementsByPool, projectId));
+    // and last but not least also the single sample measurements
+    singleMeasurements.stream()
+        .map(singleMeasurement -> build(List.of(singleMeasurement), projectId))
+        .forEach(proteomicsMeasurementsMapping::putAll);
+
     return measurementDomainService.addProteomicsAll(proteomicsMeasurementsMapping);
   }
 
-  private Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> preparePxpMeasurement(
-      ProjectId projectId, ProteomicsMeasurementMetadata metadata) {
-    Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> proteomicsMeasurements = new HashMap<>();
-    var associatedSampleCodes = metadata.associatedSamples();
-    var selectedSampleCode = MeasurementCode.createMS(
-        String.valueOf(metadata.associatedSamples().get(0).code()));
-    var sampleIdCodeEntries = queryIdCodePairs(associatedSampleCodes);
-
-    if (sampleIdCodeEntries.size() != associatedSampleCodes.size()) {
-      log.error(
-          "Could not find all corresponding sample ids for input: " + associatedSampleCodes);
-      throw new MeasurementRegistrationException(ErrorCode.FAILED);
+  /**
+   * Merges and builds {@link ProteomicsMeasurement} based on the given pool information.
+   *
+   * @param groupedMetadata already grouped measurement metadata by pool
+   * @param projectId       the project the measurement belongs to
+   * @return
+   * @since 1.0.0
+   */
+  private Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> mergeByPool(
+      Map<String, List<ProteomicsMeasurementMetadata>> groupedMetadata, ProjectId projectId) {
+    Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> metadataMap = new HashMap<>();
+    for (String poolLabel : groupedMetadata.keySet()) {
```

Comment:
As you are using both the key and value, please iterate over the entrySet instead and access entry.key() and entry.value().

Reply:
Can you explain, what the issue is with the current implementation and what your's does improve? Otherwise I consider it as different flavor and thus as not relevant for the PR

Reply:
Here you loop over a map. When looping over the keys and getting the value in every iteration, you access the map 2 times. When iterating over the entries and extracting key and value from that you only loop once. My comment is in response to a Sonarlint recommendation for performance. As I did not think that this needs a big change I thought let's take all the performance we can get 😄 
Feel free to leave it like this as I do not think that the map will be to big and I do not have clear data on performance changes.